src/poseidon.c

```diff
@@ -0,0 +1,1388 @@
+      0x00, 0x00, 0x00, 0x00, 0xfd, 0xa4, 0x35, 0x30, 0x8c, 0x29, 0x81, 0x30,
+      0xfb, 0x6f, 0xee, 0xd9, 0xb6, 0x46, 0xaa, 0x18, 0x4a, 0x17, 0xe1, 0x1d,
+      0xa6, 0x84, 0x48, 0xa3, 0xe5, 0x98, 0x1e, 0x22, 0xff, 0x1a, 0x63, 0x5b},
+     {0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+      0x00, 0x00, 0x00, 0x00, 0x48, 0xe3, 0x32, 0xe3, 0xff, 0x61, 0x23, 0x59,
+      0x04, 0x3b, 0x78, 0xcb, 0x7a, 0x8e, 0xc2, 0x6c, 0x14, 0x03, 0x79, 0xf5,
+      0xef, 0xc4, 0x2d, 0x92, 0x8e, 0xba, 0x02, 0x45, 0x42, 0xba, 0xb8, 0x80},
+     {0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+      0x00, 0x00, 0x00, 0x00, 0xeb, 0x2c, 0xaa, 0x98, 0xe5, 0x48, 0xca, 0x69,
+      0x3b, 0xed, 0xe6, 0x43, 0x0d, 0xcf, 0x30, 0xca, 0x6d, 0xac, 0xba, 0xe2,
+      0xdb, 0xc1, 0x54, 0x68, 0x67, 0xb6, 0xfb, 0xd7, 0xfb, 0xf8, 0xb7, 0x15}},
+    {{0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+      0x00, 0x00, 0x00, 0x00, 0x5c, 0xd9, 0x35, 0x56, 0x2f, 0x3e, 0x6f, 0x8a,
+      0x3f, 0x64, 0xb4, 0xd2, 0x9a, 0x5f, 0xdb, 0xd8, 0x41, 0x7c, 0xd9, 0x45,
+      0x92, 0x98, 0xf9, 0xbe, 0x0a, 0x7e, 0x45, 0x70, 0x59, 0xd5, 0x6b, 0x5c},
+     {0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+      0x00, 0x00, 0x00, 0x00, 0xea, 0x59, 0x6a, 0x68, 0x71, 0x46, 0xc4, 0x25,
+      0x8a, 0x70, 0xf7, 0x0c, 0x08, 0xac, 0xe8, 0xd7, 0x22, 0xbf, 0xca, 0xe3,
+      0xcc, 0x0f, 0xe4, 0xcb, 0xd3, 0x10, 0x51, 0xd4, 0xb7, 0x82, 0x44, 0x4c},
+     {0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+      0x00, 0x00, 0x00, 0x00, 0x24, 0xc7, 0xf0, 0xf4, 0x96, 0xf1, 0x89, 0x94,
+      0xa9, 0x13, 0xd3, 0x1a, 0x85, 0x8d, 0x1d, 0x6f, 0x7d, 0x42, 0x24, 0xa9,
+      0xcb, 0x4a, 0xda, 0x9e, 0x5e, 0xf2, 0xd8, 0x59, 0x16, 0x31, 0x1a, 0xb9}},
+    {{0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+      0x00, 0x00, 0x00, 0x00, 0x60, 0x1a, 0x4c, 0x5d, 0x5f, 0x8d, 0x67, 0x59,
+      0x16, 0x84, 0x69, 0xbe, 0x48, 0x67, 0xec, 0x6d, 0xb2, 0x9c, 0x81, 0x4a,
+      0x85, 0x77, 0xba, 0x76, 0xea, 0xbe, 0x08, 0x2a, 0xce, 0x0d, 0x25, 0xe0},
+     {0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+      0x00, 0x00, 0x00, 0x00, 0xe2, 0x1c, 0x67, 0x1a, 0x66, 0x7f, 0x6c, 0xfd,
+      0xf1, 0xad, 0xdd, 0x34, 0x53, 0x22, 0x4b, 0xea, 0xbc, 0x02, 0x43, 0x94,
+      0xeb, 0x59, 0xc1, 0x73, 0x78, 0x14, 0x3b, 0xa1, 0x29, 0xb8, 0x49, 0x45},
+     {0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+      0x00, 0x00, 0x00, 0x00, 0x01, 0xc3, 0x4e, 0x11, 0x60, 0x7c, 0xd0, 0x7d,
+      0x27, 0xb3, 0x50, 0x78, 0x4b, 0x2a, 0xd5, 0x14, 0x84, 0x2a, 0x39, 0x2e,
+      0xb3, 0x53, 0x82, 0x5c, 0x65, 0xce, 0x7b, 0xda, 0xc1, 0x1c, 0xc2, 0x74}},
+    {{0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+      0x00, 0x00, 0x00, 0x00, 0xd5, 0xe3, 0xb0, 0x15, 0x53, 0x04, 0x96, 0x47,
+      0xef, 0x77, 0x4d, 0x4c, 0x5
```

Comment:
this is undefined behavior: t0 is a destination and a source of this function, but the source argument is marked const. const requires there be no other mutations. the code might work as written (being careful to read all source bytes before writing any dest bytes) but compiler optimizations might break it anyway. we can either rewrite the code to avoid the aliases, compile with -fno-strict-aliasing, or remove the const.  
there are several other aliasing problems in the lines below.

Reply:
iiuc `const` here means that the function `madd(c, const a, const b)` (which is being called by `field_add(c, const a, const b)` ) doesn't alter `a` or `b`, but there's nothing saying that `c` and `a` can't be pointing to the same memory. `memmove` has a similar signature for example, `void *memmove(void *dest, const void *src, size_t n);)`, but its `src` and `dest` can explicitly be overlapping without issue

Reply:
I'm pretty sure you're right, strict aliasing here doesn't matter because the pointers have the same type so they'll always be assumed to possibly alias unless `restrict` is added.